<a href="https://colab.research.google.com/github/SantiagOXO/math/blob/main/practicas/practica_combinatoria.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica 1. Combinatoria
Resuelva los siguientes apartados de la práctica.  
La práctica 1.1 se proporciona solucionada para que pueda servir como ejemplo.


## Práctica 1.1
Cuatro amigos:
* Antonio (A)
* Begoña (B)
* Cesar (C)
* Diana (D)

constituyen una asociación y tienen que asignar los siguientes cargos:
* Presiente (1)
* Vicepresidente (2)
* Secretario (3)
* Tesorero (4)

Se pide:
1. Escribiendo la solución en una celda de Markdown analice el caso indicando si se trata de variaciones, permutaciones o combinaciones, con o sin repetición o el caso que sea, y el motivo para clasificar el caso de esta forma.
2. Utilizando fórmulas de $\LaTeX$ calcule cuántos casos existen.
3. Cree un programa en Python para calcule cuántos casos existen.
4. Cree un programa en Python para imprimir todas las posibilidades y numérelas. Por ejemplo:
    - 1 ABCD
    - 2 ABDC
    - 3 ACBD
    - &#20; &#20; &#20; ... ... ...

### Solución 1.1
* Estamos ante un caso de **permutaciones sin repetición** ya que:
 - El orden importa, no es lo mismo estar el nº1 que significa ser presidente que el nº 4 que corresponde al tesorero.
 - Intervienen todos. Son 4 amigos para 4 cargos, todos tienen cargo, lo que cambiar es el orden, esto es, el cargo que obtengan.
 - No se repiten elementos. Una persona no puede tener dos cargos simultaneamente.
* Los casos que se obtienen con las permutaciones sin repetición son $P_{n}=n!$ que para este ejemplo son:

$$P_{4}=4!=4 \cdot 3 \cdot 2 \cdot 1 = 24$$

In [2]:
import math
p = math.factorial(4)
print("Los casos son: ", p)
print("Comprobación:  ", 4*3*2*1)

Los casos son:  24
Comprobación:   24


#### Método 1
Sin usar librería.

In [ ]:
lista = ['A', 'B', 'C', 'D']
contador = 1

for i in lista:
    for j in lista:
        for k in lista:
            for l in lista:
                if i != j and i != k and i != l and j != k and j != l and k != l:
                    print(f"{contador:2d} →  {i+j+k+l}")
                    contador += 1

#### Método 2
* Usamos la función `permutations` de la librería `itertools`.
* Imprimimos todas las permutaciones de los elementos de una lista.

In [ ]:
from itertools import permutations

perm = permutations(['A','B','C','D'])   # obtenemos todas las permutaciones de la lista 

for i in list(perm):                     # imprimimos todas las permutaciones 
    print (*i)                           # con el asterisco se muestran las listas sin corchetes ni comas

## Práctica 1.2
Cuatro amigos:
* Antonio (A)
* Begoña (B)
* Cesar (C)
* Diana (D)

constituyen una asociación y tienen que asignar los siguientes cargos:
* Presiente (1)
* Secretario (3)
* Tesorero (4)

Se pide:
1. Escribiendo la solución en una celda de Markdown analice el caso indicando si se trata de variaciones, permutaciones o combinaciones, con o sin repetición o el caso que sea, y el motivo para clasificar el caso de esta forma.
2. Utilizando fórmulas de $\LaTeX$ calcule cuántos casos existen.
3. Cree un programa en Python para calcule cuántos casos existen.
4. Cree un programa en Python para imprimir todas las posibilidades y numérelas. Por ejemplo:
    - 1 ABC
    - 2 ABD
    - 3 ACB
    - &#20; &#20; &#20; ... ... ...

### Solución 1.2

### Es una variación sin repetición
El orden si importa, se toman algunos elementos (3) y no se repiten los mismos

$$
V_{n, k} = n * (n-1) ... (n-k+1) = \frac{n!}{(n-k)!} = \frac{4!}{(4-3)!} = 24
$$

In [1]:
import math

print ("Las variaciones de 4 tomadas de 3 en 3 son: : ", end="")
print (math.factorial(4)/math.factorial(4-3))
print("Comprobación: 4*3*2*1 =",4*3*2*1)

Las variaciones de 4 tomadas de 3 en 3 son: : 24.0
Comprobación: 4*3*2*1 = 24


In [ ]:
import math

print ("Las variaciones de 4 elementos tomados de 3 en 3 son: : ", end="")
print (math.factorial(4)/math.factorial(4-3))d
print("Comprobación: 4*3*2*1=",4*3*2*1)
cargos = "ABCD"
contador = 0
for i in cargos:
    for j in cargos:
        for k in cargos:
            if i != j and i != k and j !=k:
                contador += 1
                print(f"{contador:2d} {i+j+k}")

Las variaciones de 4 elementos tomados de 3 en 3 son: : 24.0
Comprobación: 4*3*2*1= 24
 1 ABC
 2 ABD
 3 ACB
 4 ACD
 5 ADB
 6 ADC
 7 BAC
 8 BAD
 9 BCA
10 BCD
11 BDA
12 BDC
13 CAB
14 CAD
15 CBA
16 CBD
17 CDA
18 CDB
19 DAB
20 DAC
21 DBA
22 DBC
23 DCA
24 DCB


## Práctica 1.3
Invente un caso de **variaciones sin repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.3

#### 7 jugadores de baloncesto son elegidos por 5 equipos antes de la temporada, ¿cuántas posibilidades existen?

$$
V_{n, k} = n * (n-1) ... (n-k+1) = \frac{n!}{(n-k)!} = \frac{7!}{(7-5)!} = 2520
$$

### Estamos ante una variación sin repetición
El orden si importa, se toman algunos elementos (5) y no se pueden repetir los elementos

In [4]:
import math

print ("Las variaciones de 7 tomadas de 5 en 5 son: : ", end="")
print (math.factorial(7)/math.factorial(7-5))
print("Comprobación: 7*6*5*4*3 =",7*6*5*4*3)

Las variaciones de 7 tomadas de 5 en 5 son: : 2520.0
Comprobación: 7*6*5*4*3 = 2520


In [17]:
import math

print ("Las variaciones de 7 tomadas de 5 en 5 son: : ", end="")
print (math.factorial(7)/math.factorial(7-5))
print("Comprobación: 7*6*5*4*3 =",7*6*5*4*3)
jugadores = "ABCDEFG"
contador = 0
for i in jugadores:
    for j in jugadores:
        for k in jugadores:
            for l in jugadores:
                for m in jugadores:
                    if i !=j and i != k and i != l and i != m and j != k and j != l and j != m and k != l and k != m and l != m:
                        contador += 1
                        print(f"{contador:2d} {i+j+k+l+m}")


Las variaciones de 7 tomadas de 5 en 5 son: : 2520.0
Comprobación: 7*6*5*4*3 = 2520
 1 ABCDE
 2 ABCDF
 3 ABCDG
 4 ABCED
 5 ABCEF
 6 ABCEG
 7 ABCFD
 8 ABCFE
 9 ABCFG
10 ABCGD
11 ABCGE
12 ABCGF
13 ABDCE
14 ABDCF
15 ABDCG
16 ABDEC
17 ABDEF
18 ABDEG
19 ABDFC
20 ABDFE
21 ABDFG
22 ABDGC
23 ABDGE
24 ABDGF
25 ABECD
26 ABECF
27 ABECG
28 ABEDC
29 ABEDF
30 ABEDG
31 ABEFC
32 ABEFD
33 ABEFG
34 ABEGC
35 ABEGD
36 ABEGF
37 ABFCD
38 ABFCE
39 ABFCG
40 ABFDC
41 ABFDE
42 ABFDG
43 ABFEC
44 ABFED
45 ABFEG
46 ABFGC
47 ABFGD
48 ABFGE
49 ABGCD
50 ABGCE
51 ABGCF
52 ABGDC
53 ABGDE
54 ABGDF
55 ABGEC
56 ABGED
57 ABGEF
58 ABGFC
59 ABGFD
60 ABGFE
61 ACBDE
62 ACBDF
63 ACBDG
64 ACBED
65 ACBEF
66 ACBEG
67 ACBFD
68 ACBFE
69 ACBFG
70 ACBGD
71 ACBGE
72 ACBGF
73 ACDBE
74 ACDBF
75 ACDBG
76 ACDEB
77 ACDEF
78 ACDEG
79 ACDFB
80 ACDFE
81 ACDFG
82 ACDGB
83 ACDGE
84 ACDGF
85 ACEBD
86 ACEBF
87 ACEBG
88 ACEDB
89 ACEDF
90 ACEDG
91 ACEFB
92 ACEFD
93 ACEFG
94 ACEGB
95 ACEGD
96 ACEGF
97 ACFBD
98 ACFBE
99 ACFBG
100 ACFDB
101 ACFDE
102 A

## Práctica 1.4
Invente un caso de **variaciones con repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.4

### Con las cifras del 1 al 5, ¿cuántos números de 3 cifras se pueden formar?

$$
VR _{n, k} = n^k = 5^3 = 125
$$

### Estamos ante una variación con repetición
El orden si importa, no se cuentan con todos los elementos (3), y se pueden repetir los elementos

In [12]:
import math

print ("Las variaciones con repetición de 5 tomadas de 3 en 3 son: : ", end="")
print (pow(5,3))
print("Comprobación: 5*5*5 =",5*5*5)



Las variaciones con repetición de 5 tomadas de 3 en 3 son: : 125
Comprobación: 5*5*5 = 125


In [1]:
import math

print ("Las variaciones con repetición de 5 tomadas de 3 en 3 son: : ", end="")
print (pow(5,3))
print("Comprobación: 5*5*5 =",5*5*5)

cifras = "12345"
contador = 0
for i in cifras:
    for j in cifras:
        for k in cifras:
                contador += 1
                print(f"{contador:2d} {i+j+k}")

Las variaciones con repetición de 5 tomadas de 3 en 3 son: : 125
Comprobación: 5*5*5 = 125
 1 111
 2 112
 3 113
 4 114
 5 115
 6 121
 7 122
 8 123
 9 124
10 125
11 131
12 132
13 133
14 134
15 135
16 141
17 142
18 143
19 144
20 145
21 151
22 152
23 153
24 154
25 155
26 211
27 212
28 213
29 214
30 215
31 221
32 222
33 223
34 224
35 225
36 231
37 232
38 233
39 234
40 235
41 241
42 242
43 243
44 244
45 245
46 251
47 252
48 253
49 254
50 255
51 311
52 312
53 313
54 314
55 315
56 321
57 322
58 323
59 324
60 325
61 331
62 332
63 333
64 334
65 335
66 341
67 342
68 343
69 344
70 345
71 351
72 352
73 353
74 354
75 355
76 411
77 412
78 413
79 414
80 415
81 421
82 422
83 423
84 424
85 425
86 431
87 432
88 433
89 434
90 435
91 441
92 442
93 443
94 444
95 445
96 451
97 452
98 453
99 454
100 455
101 511
102 512
103 513
104 514
105 515
106 521
107 522
108 523
109 524
110 525
111 531
112 532
113 533
114 534
115 535
116 541
117 542
118 543
119 544
120 545
121 551
122 552
123 553
124 554
125 555


## Práctica 1.5
Invente un caso de **permutaciones sin repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.5

### ¿Cuántos números de 5 cifras diferentes se pueden formar con los números del 1 al 5?

$$
P_N = N! = 5! = 120
$$

### Estamos ante un caso de permutación sin repetición
El orden si importa, se cuentan todos los elementos (5), y no se pueden repetir

In [15]:
import math

print ("La permutación de 5 cifras es igual a: : ", end="")
print (math.factorial(5))
print("Comprobación: 5*4*3*2*1 =",5*4*3*2*1)


La permutación de 5 cifras es igual a: : 120
Comprobación: 5*4*3*2*1 = 120


In [2]:
import math

print ("Los números que se pueden formar son: : ", end="")
print (math.factorial(5))
print("Comprobación: 5*4*3*2*1 =",5*4*3*2*1)
cifras = "12345"
contador = 0
for i in cifras:
    for j in cifras:
        for k in cifras:
            for l in cifras:
                for m in cifras:
                    if i !=j and i != k and i != l and i != m and j != k and j != l and j != m and k != l and k != m and l != m:
                        contador += 1
                        print(f"{contador:2d} {i+j+k+l+m}")





Los números que se pueden formar son: : 120
Comprobación: 5*4*3*2*1 = 120
 1 12345
 2 12354
 3 12435
 4 12453
 5 12534
 6 12543
 7 13245
 8 13254
 9 13425
10 13452
11 13524
12 13542
13 14235
14 14253
15 14325
16 14352
17 14523
18 14532
19 15234
20 15243
21 15324
22 15342
23 15423
24 15432
25 21345
26 21354
27 21435
28 21453
29 21534
30 21543
31 23145
32 23154
33 23415
34 23451
35 23514
36 23541
37 24135
38 24153
39 24315
40 24351
41 24513
42 24531
43 25134
44 25143
45 25314
46 25341
47 25413
48 25431
49 31245
50 31254
51 31425
52 31452
53 31524
54 31542
55 32145
56 32154
57 32415
58 32451
59 32514
60 32541
61 34125
62 34152
63 34215
64 34251
65 34512
66 34521
67 35124
68 35142
69 35214
70 35241
71 35412
72 35421
73 41235
74 41253
75 41325
76 41352
77 41523
78 41532
79 42135
80 42153
81 42315
82 42351
83 42513
84 42531
85 43125
86 43152
87 43215
88 43251
89 43512
90 43521
91 45123
92 45132
93 45213
94 45231
95 45312
96 45321
97 51234
98 51243
99 51324
100 51342
101 51423
102 51432
103 5

## Práctica 1.6
Invente un caso de **permutaciones con repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.6

### 4 deportistas de élite compiten en 4 deportes diferentes, y los 4 pueden ganar en más de un deporte, ¿Cuántas combinaciones posibles hay?

Atleta 1 (1)
Atleta 2 (2)
Atleta 3 (3)
Atleta 4 (4)

$$
PR_{n, k} = \frac{n!}{a!*b!*c!...} 
$$

### Estamos frente a una permutación con repetición:
El orden si importa, se toman todos los elementos y se repiten elementos

In [54]:
lista = ['1 ', '2 ', '3 ', '4 ']
contador = 0
print("El total de posibilidades es 256")
print("Comprobación= 4^4 =",256)
for i in lista:
    for j in lista:
        for k in lista:
            for l in lista:
                contador += 1
                print(f"{contador:2d} {i+j+k+l}")


El total de posibilidades es 256
Comprobación= 4^4 = 256
 1 1 1 1 1 
 2 1 1 1 2 
 3 1 1 1 3 
 4 1 1 1 4 
 5 1 1 2 1 
 6 1 1 2 2 
 7 1 1 2 3 
 8 1 1 2 4 
 9 1 1 3 1 
10 1 1 3 2 
11 1 1 3 3 
12 1 1 3 4 
13 1 1 4 1 
14 1 1 4 2 
15 1 1 4 3 
16 1 1 4 4 
17 1 2 1 1 
18 1 2 1 2 
19 1 2 1 3 
20 1 2 1 4 
21 1 2 2 1 
22 1 2 2 2 
23 1 2 2 3 
24 1 2 2 4 
25 1 2 3 1 
26 1 2 3 2 
27 1 2 3 3 
28 1 2 3 4 
29 1 2 4 1 
30 1 2 4 2 
31 1 2 4 3 
32 1 2 4 4 
33 1 3 1 1 
34 1 3 1 2 
35 1 3 1 3 
36 1 3 1 4 
37 1 3 2 1 
38 1 3 2 2 
39 1 3 2 3 
40 1 3 2 4 
41 1 3 3 1 
42 1 3 3 2 
43 1 3 3 3 
44 1 3 3 4 
45 1 3 4 1 
46 1 3 4 2 
47 1 3 4 3 
48 1 3 4 4 
49 1 4 1 1 
50 1 4 1 2 
51 1 4 1 3 
52 1 4 1 4 
53 1 4 2 1 
54 1 4 2 2 
55 1 4 2 3 
56 1 4 2 4 
57 1 4 3 1 
58 1 4 3 2 
59 1 4 3 3 
60 1 4 3 4 
61 1 4 4 1 
62 1 4 4 2 
63 1 4 4 3 
64 1 4 4 4 
65 2 1 1 1 
66 2 1 1 2 
67 2 1 1 3 
68 2 1 1 4 
69 2 1 2 1 
70 2 1 2 2 
71 2 1 2 3 
72 2 1 2 4 
73 2 1 3 1 
74 2 1 3 2 
75 2 1 3 3 
76 2 1 3 4 
77 2 1 4 1 
78 2 1 4 2 
79 2 1 

## Práctica 1.7
Invente un caso de **combinaciones sin repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.7

### De entre 5 tipos de helado, un niño tiene la posibilidad de escoger 2 y no puede escoger los mismos


*   Fresa(F)
*   Arándanos(A)
*   Menta(M)
*   Pistacho(P)
*   Chocolate(C)



$$
C_{n, k} = \frac{n!}{n!(n-k)!} = \frac{5!}{2!*(5-2)!} = 10
$$

### Estamos ante un caso de combinaciones sin repetición:
El orden no importa, no intervienen todos. Son 5 sabores y solo 2 pueden ser elegidos.
No se repiten elementos. No puede escoger el mismo sabor dos veces.

In [30]:
from itertools import combinations
c = combinations('FAMPV', 2)
print("El número de combinaciones posibles es 15")
print("Comprobación = 6*5*4*3*2*1/48 =", 6*5*4*3*2*1/48)
contador = 0
for i in c:
  contador += 1
  print(f"{contador:2d} {i}")

El número de combinaciones posibles es 15
Comprobación = 6*5*4*3*2*1/48 = 15.0
 1 ('F', 'A')
 2 ('F', 'M')
 3 ('F', 'P')
 4 ('F', 'V')
 5 ('A', 'M')
 6 ('A', 'P')
 7 ('A', 'V')
 8 ('M', 'P')
 9 ('M', 'V')
10 ('P', 'V')


## Práctica 1.8
Invente un caso de **combinaciones con repetición**.
Se pide contestar a los mismos puntos que el caso anterior.

### Solución 1.8

### De entre 5 tipos de helado, un niño tiene la posibilidad de escoger 2 con la posibilidad de repetir sabor

*   Fresa(F)
*   Arándanos(A)
*   Menta(M)
*   Pistacho(P)
*   Chocolate(C)



$$CR^n_m = CR_{n,m} = C_{n,m+n-1} = \binom{m+n-1}{n} = \frac{(m+n-1)!}{n! \cdot (m-1)!} $$

### Estamos ante una combinación con repetición
El orden no importa, no intervienen todos. Son 5 sabores y solo 2 pueden ser elegidos. Esta vez si se repiten elementos. La persona puede escoger dos veces el mismo sabor.

In [31]:
from itertools import combinations_with_replacement
c = combinations_with_replacement('FAMPC', 2)
m = 5
n = 2
print(f" Los casos posibles son:{int(math.factorial(m+n-1)/math.factorial(n)/math.factorial(m-1))}")
print("Comprobación= 6*5*4*3*2*1/48 =", 6*5*4*3*2*1/48)
contador = 0
for i in c:
    contador += 1
    print(f"{contador:2d} {i}")



 Los casos posibles son:15
Comprobación= 6*5*4*3*2*1/48 = 15.0
 1 ('F', 'F')
 2 ('F', 'A')
 3 ('F', 'M')
 4 ('F', 'P')
 5 ('F', 'C')
 6 ('A', 'A')
 7 ('A', 'M')
 8 ('A', 'P')
 9 ('A', 'C')
10 ('M', 'M')
11 ('M', 'P')
12 ('M', 'C')
13 ('P', 'P')
14 ('P', 'C')
15 ('C', 'C')


## Práctica 1.9
En el gobierno se crea una unidad de crisis formada por el presidente (P) y 4 ministros (A,B,C,D) que se reúnen sentados alrededor de una **mesa redonda**.

Se pide:  
1. Calcule de cuantas formas pueden llegar a sentarse si el ministro de defensa (D) siempre ha de estar al lado del presidente (P). Escriba la solución en una celda de Markdown utilizando fórmulas en $\LaTeX$  explicando el caso.
2. Cree un programa en Python para calcular el número de casos que se pueden dar al sentarse en la mesa.
3. Cree un programa en Python para imprimir todas las posibilidades y numérelas. Por ejemplo:
    - 1 PDABC
    - 2 PDACB
    - 3 PDBCD
    - 4 PDBDC
    - 5 PDCAB
    - 6 PDCBA
    - 7 DPABC
    - 8 DPACB
    - &#20; &#20; &#20; ... ... ...


### Solución 1.9

## Práctica 1.10
* Realice una breve explicación teórica del **triángulo de Pascal**, o también llamado triángulo de Tartaglia, y de las fórmulas de los números combinatorios.
* Cree un programa en Python para hacer el triángulo de Pascal.

### Solución 1.10

El triángulo de Pascal se define como un tipo de patrón numérico en el que los números se organizan para que parezcan un triángulo. En este concepto de Matemáticas se forma un array triangular, formada por números que son la suma de la fila adyacente. Además, los bordes externos son siempre 1.



### La fórmula que permite hallar el valor de un número combinatorio es la siguiente:

$$
C_{n, k} = \frac{n!}{n!(n-k)!} 
$$

###Y cuándo se repiten:

$$CR^n_m = CR_{n,m} = C_{n,m+n-1} = \binom{m+n-1}{n} = \frac{(m+n-1)!}{n! \cdot (m-1)!} $$

In [33]:
num = int(input("Enter the number of rows:"))

for n in range(num):
    print(' '*(num-n), end='')

    print(' '.join(map(str, str(11**n))))

Enter the number of rows:39
                                       1
                                      1 1
                                     1 2 1
                                    1 3 3 1
                                   1 4 6 4 1
                                  1 6 1 0 5 1
                                 1 7 7 1 5 6 1
                                1 9 4 8 7 1 7 1
                               2 1 4 3 5 8 8 8 1
                              2 3 5 7 9 4 7 6 9 1
                             2 5 9 3 7 4 2 4 6 0 1
                            2 8 5 3 1 1 6 7 0 6 1 1
                           3 1 3 8 4 2 8 3 7 6 7 2 1
                          3 4 5 2 2 7 1 2 1 4 3 9 3 1
                         3 7 9 7 4 9 8 3 3 5 8 3 2 4 1
                        4 1 7 7 2 4 8 1 6 9 4 1 5 6 5 1
                       4 5 9 4 9 7 2 9 8 6 3 5 7 2 1 6 1
                      5 0 5 4 4 7 0 2 8 4 9 9 2 9 3 7 7 1
                     5 5 5 9 9 1 7 3 1 3 4 9 2 2 3 1 4 8 1
                    6 